independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java

```diff
@@ -149,6 +159,42 @@ private void implementUnlockAll(ClassCreator contextInstances, Map<String, Insta
         unlockAll.returnVoid();
     }
 
+    private void implementCheckFirstRemoveEach(ClassCreator contextInstances, Map<String, InstanceAndLock> idToFields) {
+        MethodCreator removeEach = contextInstances.getMethodCreator("removeEach", void.class, Consumer.class)
+                .setModifiers(ACC_PUBLIC);
+        // ContextInstanceHandle<?> copy1 = this.1;
+        // if (copy1 != null) {
+        //    this.1l.lock();
+        //    // we don't really trust that things haven't changed
+        //    copy1 = this.1;
+        //    this.1 = null;
+        //    this.1l.unlock();
+        //  }
+        //  if (action != null)
+        //    if (copy1 != null) {
+        //       consumer.accept(copy1);
+        //    }
+        // }
+        List<ResultHandle> results = new ArrayList<>(idToFields.size());
+        for (InstanceAndLock fields : idToFields.values()) {
+            AssignableResultHandle copy = removeEach.createVariable(ContextInstanceHandle.class);
+            removeEach.assign(copy, removeEach.readInstanceField(fields.instance, removeEach.getThis()));
+            results.add(copy);
+            BytecodeCreator isNotNull = removeEach.ifNotNull(copy).trueBranch();
+            ResultHandle lock = isNotNull.readInstanceField(fields.lock, isNotNull.getThis());
+            isNotNull.invokeInterfaceMethod(MethodDescriptors.LOCK_LOCK, lock);
+            isNotNull.assign(copy, isNotNull.readInstanceField(fields.instance, isNotNull.getThis()));
+            isNotNull.writeInstanceField(fields.instance, isNotNull.getThis(), isNotNull.loadNull());
+            isNotNull.invokeInterfaceMethod(MethodDescriptors.LOCK_UNLOCK, lock);
+        }
+        BytecodeCreator consumerIsNotNull = removeEach.ifNotNull(removeEach.getMethodParam(0)).trueBranch();
+        for (ResultHandle result : results) {
+            consumerIsNotNull.ifNotNull(result).trueBranch().invokeInterfaceMethod(MethodDescriptors.CONSUMER_ACCEPT,
+                    removeEach.getMethodParam(0), result);
```

Comment:
This should be

```suggestion
                    consumerIsNotNull.getMethodParam(0), result);
```

but it probably doesn't matter.